### ISSUE TITLE
Use a fixed seed in SLiM engine tests.

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/stdpopsim#1162]
Maybe this is an issue with seed propagation, but not having a fixed seed can lead to intermittent test failures, e.g. 

```python
__________________________ TestCLI.test_simulate_dfe ___________________________

self = <tests.test_slim_engine.TestCLI object at 0x7fdc0d89e450>

    def test_simulate_dfe(self):
        slim_path = os.environ.get("SLIM", "slim")
    
        def count_mut_types(ts):
            mut_info = {}
            for mut in ts.mutations():
                for j, md in zip(
                    mut.derived_state.split(","), mut.metadata["mutation_list"]
                ):
                    if j not in mut_info:
                        mut_info[int(j)] = md
            num_neutral = sum([mut_info[j]["selection_coeff"] == 0.0 for j in mut_info])
            return [num_neutral, abs(len(mut_info) - num_neutral)]
    
        # test a DFE without demography
        with tempfile.NamedTemporaryFile(mode="w") as f:
            cmd = (
                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
                f"HomSap -c chr22 -l0.1 -o {f.name} --dfe Gamma_K17 10"
            ).split()
            capture_output(stdpopsim.cli.stdpopsim_main, cmd)
            ts = tskit.load(f.name)
        assert ts.num_samples == 10
        assert all(tree.num_roots == 1 for tree in ts.trees())
        n_mut_types = count_mut_types(ts)
        assert n_mut_types[0] > 0
        assert n_mut_types[1] > 0
    
        # test a DFE without demography with interval
        with tempfile.NamedTemporaryFile(mode="w") as f:
            cmd = (
                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
                f"HomSap -c chr22 -l0.01 -o {f.name} --dfe Gamma_K17 "
                f"--dfe-interval 1000,100000 10"
            ).split()
            capture_output(stdpopsim.cli.stdpopsim_main, cmd)
            ts = tskit.load(f.name)
        assert ts.num_samples == 10
        assert all(tree.num_roots == 1 for tree in ts.trees())
        n_mut_types = count_mut_types(ts)
        assert n_mut_types[0] > 0
        assert n_mut_types[1] > 0
    
        # test DFE with demography and interval
        with tempfile.NamedTemporaryFile(mode="w") as f:
            cmd = (
                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
                f"HomSap -c chr22 -l0.01 -o {f.name} "
                "-d OutOfAfrica_3G09 --dfe Gamma_K17 "
                "--dfe-interval 1000,100000 10"
            ).split()
            capture_output(stdpopsim.cli.stdpopsim_main, cmd)
            ts = tskit.load(f.name)
        assert ts.num_samples == 10
        assert all(tree.num_roots == 1 for tree in ts.trees())
        n_mut_types = count_mut_types(ts)
        assert n_mut_types[0] > 0
        assert n_mut_types[1] > 0
    
        # test DFE with dfe-annotation
        with tempfile.NamedTemporaryFile(mode="w") as f:
            cmd = (
                f"-q -e slim --slim-scaling-factor 20 --slim-path {slim_path} "
                f"HomSap -c chr22 -o {f.name} "
                "-d OutOfAfrica_3G09 --dfe Gamma_K17 "
                "--dfe-annotation ensembl_havana_104_CDS 10"
            ).split()
>           capture_output(stdpopsim.cli.stdpopsim_main, cmd)

tests/test_slim_engine.py:377: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_cli.py:41: in capture_output
    func(*args, **kwargs)
stdpopsim/cli.py:1068: in stdpopsim_main
    run(args)
stdpopsim/cli.py:1059: in run
    args.runner(args)
stdpopsim/cli.py:803: in run_simulation
    ts = engine.simulate(**kwargs)
stdpopsim/slim_engine.py:1284: in simulate
    verbosity=verbosity,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = _SLiMEngine(), script_file = '/tmp/tmp9gpw1mxo.slim', slim_path = 'slim'
seed = None, dry_run = False, verbosity = None

    def _run_slim(
        self, script_file, slim_path=None, seed=None, dry_run=False, verbosity=None
    ):
        """
        Run SLiM.
    
        We capture the output using Popen's line-oriented text buffering
        (bufsize=1, universal_newlines=True) and redirect all messages to
        Python's logging module.
        By convention, messages from SLiM prefixed with "ERROR: " or
        "WARNING: " are treated as ERROR or WARN loglevels respectively.
        All other output on stdout is given the DEBUG loglevel.
        ERROR messages will raise a SLiMException here too, because
        they are always generated by the `stop()` eidos function which
        makes SLiM exit with a non-zero return code.
        """
        if slim_path is None:
            slim_path = self.slim_path()
    
        # SLiM v3.6 sends `stop()` output to stderr, which we rely upon.
        self._assert_min_version("3.6", slim_path)
    
        slim_cmd = [slim_path]
        if seed is not None:
            slim_cmd.extend(["-s", f"{seed}"])
        if dry_run:
            slim_cmd.extend(["-d", "dry_run=T"])
        if verbosity is not None:
            slim_cmd.extend(["-d", f"verbosity={verbosity}"])
        slim_cmd.append(script_file)
    
        with subprocess.Popen(
            slim_cmd,
            bufsize=1,
            universal_newlines=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        ) as proc:
            for line in proc.stdout:
                line = line.rstrip()
                if line.startswith("WARNING: "):
                    warnings.warn(
                        stdpopsim.UnspecifiedSLiMWarning(line[len("WARNING: ") :])
                    )
                else:
                    # filter `dbg` function calls that generate output
                    line = line.replace("dbg(self.source); ", "")
                    logger.debug(line)
    
            stderr = proc.stderr.read()
            for line in stderr.splitlines():
                if line.startswith("ERROR: "):
                    logger.error(line[len("ERROR: ") :])
    
        if proc.returncode != 0:
            raise SLiMException(
>               f"{slim_path} exited with code {proc.returncode}.\n{stderr}"
            )
E           stdpopsim.slim_engine.SLiMException: slim exited with code 1.
E           #WARNING (SLiMSim::RunInitializeCallbacks): with tree-sequence recording enabled and a non-zero mutation rate, a neutral mutation type was defined and used; this is legal, but usually undesirable, since neutral mutations can be overlaid later using the tree-sequence information.
E           ERROR (Subpopulation::UpdateFitness): total fitness of all individuals is <= 0.0.
E           
E           Error on script line 437, character 24:
E           
E                                   next;
E                                   ^^^^

stdpopsim/slim_engine.py:1359: SLiMException
```

**Comment by petrelharp** [comment]
I've tracked the cause of this error down - with a large `Q`, we can have quite a few mutations of large effect, and once there are enough of these, the maximum sum of mutation effects per individual falls below -30 or so, and then we run into floating point error, since everyone's fitness is like exp(-30), which is practically zero. To see this, run:
```
-q -e slim --slim-scaling-factor 50 --slim-script HomSap -c chr22 -o test.trees --dfe Gamma_K17 --dfe-annotation ensembl_havana_104_CDS 10 > test.slim
```
then append the following chunk to `test.slim`:
```
1: late() {
   catn(range(p0.individuals.sumOfMutationsOfType(2)));
}
```
This is `Q=50` and hits the problem after 1-5 generations. Changing `Q` to, say, 10 still shows a ratchet, but much slower (and I guess slow enough that we don't hit this problem within the time span of a simulation?).

This is clearly an issue with rescaling in general; however, we'd like to be able to run with large `Q` values for testing purposes. There may be a numerical fix in SLiM (eg a more numerically stable way of computing the fitnesses? I have not looked.). I think for the time being we can set seeds and avoid the largest values of `Q` where the problem happens most often.

**Comment by nspope** [comment]
Closed by #1169
